Sources/Build/BuildPlan.swift

```diff
@@ -0,0 +1,486 @@
+        let moduleCachePath = buildParameters.buildPath.appending(component: "ModuleCache")
+        return ["-fmodules-cache-path=" + moduleCachePath.asString]
+    }
+}
+
+/// Target description for a Swift module.
+public final class SwiftTargetDescription {
+
+    /// The module described by this target.
+    public let module: SwiftModule
+
+    /// The build parameters.
+    let buildParameters: BuildParameters
+
+    /// Path to the temporary directory for this target.
+    var tempsPath: AbsolutePath {
+        return buildParameters.buildPath.appending(component: module.c99name + ".build")
+    }
+
+    /// The objects in this target.
+    var objects: [AbsolutePath] {
+        return module.sources.relativePaths.map{ tempsPath.appending(RelativePath($0.asString + ".o")) }
+    }
+
+    /// The path to the swiftmodule file after compilation.
+    var moduleOutputPath: AbsolutePath { 
+        return buildParameters.buildPath.appending(component: module.c99name + ".swiftmodule") 
+    }
+
+    /// Any addition flags to be added. These flags are expected to be computed during build planning.
+    fileprivate var additionalFlags: [String] = []
+
+    /// Create a new target description with module and build parameters.
+    init(module: SwiftModule, buildParameters: BuildParameters) {
+        self.module = module
+        self.buildParameters = buildParameters
+    }
+
+    /// The arguments needed to compile this target.
+    public func compileArguments() -> [String] {
+        var args = [String]()
+        args += buildParameters.toolchain.swiftPlatformArgs
+        args += buildParameters.swiftCompilerFlags
+        args += optimizationArguments
+        args += ["-j\(SwiftcTool.numThreads)", "-DSWIFT_PACKAGE"]
+        args += additionalFlags
+        args += moduleCacheArgs
+        return args
+    }
+
+    /// Optimization arguments according to the build configuration.
+    private var optimizationArguments: [String] {
+        switch buildParameters.configuration {
+        case .debug:
+            return ["-Onone", "-g", "-enable-testing"]
+        case .release:
+            return ["-O"]
+        }
+    }
+
+    /// Module cache arguments.
+    private var moduleCacheArgs: [String] {
+        // FIXME: We use this hack to let swiftpm's functional test use shared cache so it doesn't become painfully slow.
+        if getenv("IS_SWIFTPM_TEST") != nil { return [] }
+        let moduleCachePath = buildParameters.buildPath.appending(component: "ModuleCache")
+        return ["-module-cache-path", moduleCachePath.asString]
+    }
+}
+
+/// The build description for a product.
+public final class ProductBuildDescription {
+
+    /// The reference to the product.
+    public let product: Product
+
+    /// The build parameters.
+    let buildParameters: BuildParameters
+
+    /// The path to the product binary produced.
+    public var binary: AbsolutePath {
+        return buildParameters.buildPath.appending(product.outname)
```

Comment:
Probably the logic to construct the actual filename of the built artefact belongs here rather than on the product, but that is something that can be refactored over time.